ENH Exposes latent mean and variance for GPCs

miguelgondu opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.
Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Fixes #22226. 


#### What does this implement/fix? Explain your changes.
Adds a `return_std_of_f` flag for binary Gaussian Process Classifiers. This value was already being computed in some of the methods, so this PR just exposes that variable and adds checks in the API to ensure this is only possible for binary (and not multiclass) classification using the Laplace approximation.

#### Any other comments?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

miguelgondu committed:
Adds a return std flag for GPCs

miguelgondu committed:
Fixes a docstring error by adding some spaces

miguelgondu committed:
Adds tests for GPC return std of f flag

miguelgondu committed:
Increases test coverage, makes tests more pythonic

miguelgondu commented:
Depending on whether the API decision, I would be happy to update this with the latest changes in `main`.

noashin commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -777,12 +850,28 @@ def predict_proba(self, X):
             X = self._validate_data(X, ensure_2d=True, dtype="numeric", reset=False)
         else:
             X = self._validate_data(X, ensure_2d=False, dtype=None, reset=False)
 
```
I would put the check whether `std_f` can be returned or not here, after the kernel tests. Than the check can be similar to the one done around lines 773:

`if self.n_classes_ > 2:`
 `    if return_std_of_f:`
`         raise ValueError(`
                `         "Returning the standard deviation of the "`
               `          "latent function f is only supported for GPCs "`
                `         "that use the Laplace Approximation."`
            `         )`
`    else:`
`        return self.base_estimator_.predict_proba(X)`
`else:`
`    return self.base_estimator_.predict_proba(`
`                X, return_std_of_f=return_std_of_f`
`            )`

miguelgondu replied:
Agreed! Will implement these changes.

noashin commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -262,18 +262,30 @@ def obj_func(theta, eval_gradient=True):
         return self
 
-    def predict(self, X):
+    def predict(self, X, return_std_of_f=False):
```
I think that `std_of_f` should be available only via `predict_proba`.
As `predict` does not provide a probabilistic estimation returning an the std of the nuiance function is out of the scope in a way.

miguelgondu replied:
Hi Noa,

First of all, thanks for the review!

I added the flag here as well to match the API for `GPR`. In the `predict` method for Gaussian Process Regression, you can also find a `return_std`. If we remove it from this `predict`, wouldn't that make the APIs different?

I'm happy to remove it either way, but I'm curious about what you think regarding whether the APIs for `GPR` and `GPC` should match.

noashin replied:
Maybe @adrinjalali could share his opinion of whether keeping the interfaces as similar as possible.

miguelgondu replied:
If it's worth anything, someone mentioned the consistency with `GPR` in a comment to the original issue: https://github.com/scikit-learn/scikit-learn/issues/22226#issuecomment-1343028703

adrinjalali replied:
I agree that in this case having it only for `predict_proba` makes more sense.

miguelgondu replied:
I'll include those changes then!

noashin reviewed (commented):
What is lacking is an explanation of the meaning of the std of f. Emphasizing that this is not the variance of \pi (the link function over the latent function) and it does not translate directly to the confidence intervals.  Maybe it can be added to the general documentation of GPs in scikit.
As the API is now more similar to the one of GaussianProcessRegressor, without further explanation, it implies that the quantity returned is equivalent to the one returned by the flag `return_std` in GaussianProcessRegressor (even though the name of the flag is different).

miguelgondu committed:
Adds documentation, moves the contribution, and addresses the review by moving the verifications

miguelgondu committed:
Adds a name for the link function in the documentation

miguelgondu commented:
Dear @noashin,

Again, many thanks for the review.

I have added a paragraph to the documentation explaining what the keyword would do, and I am hopefully clear and specific on the fact that we are returning $\sqrt{\text{Var}[f_*]}$ and not $\sqrt{\text{Var}[\pi(f)]}$. I have also moved the changelog bit to `1.4`.

Let me know if I should address anything else!

adrinjalali reviewed (commented):
I wonder if the argument could also be simply called `return_std`

miguelgondu commented:
> I wonder if the argument could also be simply called `return_std`

It might be misleading! People might assume that it refers to the uncertainty over the class probabilities $\pi(f)$ instead of the latent variable $f$. I'm happy to change it either way. What does @noashin think?

noashin commented:
> > I wonder if the argument could also be simply called `return_std`
> 
> It might be misleading! People might assume that it refers to the uncertainty over the class probabilities π(f) instead of the latent variable f. I'm happy to change it either way. What does @noashin think?

I agree. I think that return_std will be misleading.

adrinjalali commented:
ping @glemaitre maybe.

miguelgondu committed:
Removes the return std flag from the predict method

miguelgondu committed:
Updates documentation and changelog according to the change in the predict method

miguelgondu committed:
Small typo on the GPC documentation

pmeiners commented:
Any news on this feature? It seems to be requested frequently:
http://krasserm.github.io/2020/11/04/gaussian-processes-classification/
https://stackoverflow.com/questions/67818319/how-to-output-mean-and-stdv-of-gaussian-process-classifier-in-sklearn
https://stackoverflow.com/questions/67818319/how-to-output-mean-and-stdv-of-gaussian-process-classifier-in-sklearn
https://stackoverflow.com/questions/79086293/gaussian-process-binary-classification-why-is-the-variance-with-gpy-much-smalle

miguelgondu commented:
I'm happy to update it, but it seems we're waiting for @glemaitre to give input on whether we should call it `return_std` or `return_std_of_f`.

adrinjalali commented:
Ping @antoinebaker and @snath-xoc , I think you can give a nicer feedback here than I can.

antoinebaker commented:
Thanks for the PR @miguelgondu !

If I understood correctly the original issue, I see two motivations behind returning `f_std`:
1. plot the posterior latent `f` "in logit space" along with its confidence region, something like `fill_between(x, f_mean-f_std,f_mean+f_std)`
2. sample the posterior latent `f`  (from the Laplace approximation which is Gaussian with mean and covariance `f_mean` and `f_cov` ), and generate corresponding samples `y_prob = logistic(f)` to plot, use in ancestral sampling or compute bootstrap estimates (like confidence intervals or std for `y_prob`)

But if you had these applications in mind, I don't think the current approach is enough (because `f_mean` is missing) ?

I don't have a strong opinion here but maybe a separate method (name and signature TBD):
```python
laplace_approximation(X, return_std=False, return_cov=False) 
# returns f_mean, f_std or even f_cov if possible
```
 would be better that returning `f_std` inside `predict_proba`. I feel returning `y_prob` with `f_std` would mix apples with oranges and could lead to misuse (eg thinking it's `y_mean, y_std` like in the GPR case, even if the docstring warns against it).

snath-xoc commented:
Thanks for the PR @miguelgondu, I agree with @antoinebaker that taking this into a dedicated function may be better as compared to within predict_proba so we don't confuse users with y_mean and f_mean (although I see that it is well-documented). Being able to quantify the uncertainty in the latent f space could be interesting but case-specific, with the proposed "laplace_approximation" term we could also provide the option of calculating cov/std based uncertainty extimates.

miguelgondu commented:
Hi @antoinebaker and @snath-xoc,

Thanks for the feedback!

Agreed, the best path forward would be to implement another method. I'll try to do it soon, and ask for a re-review.

miguelgondu committed:
Implements a latent mean and variance function for GPC

miguelgondu committed:
Removes old contribution from v1.4

miguelgondu committed:
Adds a test on the exception raised

miguelgondu committed:
Adds a test on structured kernels

antoinebaker commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -902,3 +932,41 @@ def log_marginal_likelihood(
+        """
+        if self.n_classes_ > 2:
+            raise ValueError(
+                "Returning the mean and variance of the "
+                "latent function f is only supported for GPCs "
+                "that use the Laplace Approximation (i.e. 2 classes, "
+                f"received {self.n_classes_})."
```
```suggestion
                "Returning the mean and variance of the latent function f "
                "is only supported for binary classification, received "
                f"{self.n_classes_} classes."
```

antoinebaker reviewed (approved):
Thanks @miguelgondu for the PR, `latent_mean_and_variance` is a very good name I think. 

A few formatting nitpicks, otherwise LGTM!

miguelgondu committed:
Apply suggestions from code review regarding line wrap

Co-authored-by: antoinebaker <antoinebaker@users.noreply.github.com>

miguelgondu committed:
Addresses feedback regarding line length in docstrings

snath-xoc commented on doc/whats_new/upcoming_changes/sklearn.gaussian_process/22227.enhancement.rst:
```diff
@@ -0,0 +1 @@
+- :class:`gaussian_process.GaussianProcessClassifier` now includes a `latent_mean_and_variance` method that exposes the mean and the variance of the latent function used in the Laplace approximation. By :user:`Miguel González Duque <miguelgondu>`
```
```suggestion
- :class:`gaussian_process.GaussianProcessClassifier` now includes a `latent_mean_and_variance` method that exposes the mean and the variance of the latent function, :math:`f`, used in the Laplace approximation. By :user:`Miguel González Duque <miguelgondu>`
```

snath-xoc commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -410,6 +407,39 @@ def log_marginal_likelihood(
+
+        # Based on Algorithm 3.2 of GPML
+        K_star = self.kernel_(self.X_train_, X)  # K_star =k(x_star)
+        f_star = K_star.T.dot(self.y_train_ - self.pi_)  # Line 4
```
```suggestion
        latent_mean = K_star.T.dot(self.y_train_ - self.pi_)  # Line 4
```

snath-xoc commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -410,6 +407,39 @@ def log_marginal_likelihood(
+        # Line 6 (compute np.diag(v.T.dot(v)) via einsum)
+        var_f_star = self.kernel_.diag(X) - np.einsum("ij,ij->j", v, v)
+
+        return f_star, var_f_star
```
```suggestion
        return latent_mean, latent_var
```

miguelgondu committed:
Updates the test to reflect a change in error raised

snath-xoc commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -410,6 +407,39 @@ def log_marginal_likelihood(
+
+        Returns
+        -------
+        f_mean : array-like of shape (n_samples,)
```
```suggestion
        latent_mean : array-like of shape (n_samples,)
```

snath-xoc commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -410,6 +407,39 @@ def log_marginal_likelihood(
+        f_mean : array-like of shape (n_samples,)
+            Mean of the latent function values at the query points.
+
+        f_var : array-like of shape (n_samples,)
```
```suggestion
        latent_var : array-like of shape (n_samples,)
```

snath-xoc commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -410,6 +407,39 @@ def log_marginal_likelihood(
+        f_star = K_star.T.dot(self.y_train_ - self.pi_)  # Line 4
+        v = solve(self.L_, self.W_sr_[:, np.newaxis] * K_star)  # Line 5
+        # Line 6 (compute np.diag(v.T.dot(v)) via einsum)
+        var_f_star = self.kernel_.diag(X) - np.einsum("ij,ij->j", v, v)
```
```suggestion
        latent_var = self.kernel_.diag(X) - np.einsum("ij,ij->j", v, v)
```

snath-xoc commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -902,3 +932,40 @@ def log_marginal_likelihood(
+
+        Returns
+        -------
+        f_mean : array-like of shape (n_samples,)
```
```suggestion
        latent_mean : array-like of shape (n_samples,)
```

snath-xoc commented on sklearn/gaussian_process/_gpc.py:
```diff
@@ -902,3 +932,40 @@ def log_marginal_likelihood(
+        f_mean : array-like of shape (n_samples,)
+            Mean of the latent function values at the query points.
+
+        f_var : array-like of shape (n_samples,)
```
```suggestion
        latent_var : array-like of shape (n_samples,)
```

snath-xoc reviewed (commented):
Hi @miguelgondu thank you for the changes, I thought I would just do some nitpicks to ensure consistency in the naming of latent_mean and latent_var (unless you have a good reason for calling them f_mean and var_f_star?) throughout the code

miguelgondu committed:
Adds consistency in naming of variables

miguelgondu commented:
Hi @snath-xoc , 

Thanks for the feedback! I had left the previous names because of the notation in the book we cite. That being said, I agree with the consistency changes you proposed. They're now implemented!

Happy to address any other feedback. Let me know!

snath-xoc commented:
LGTM once we get a second opinion now!

miguelgondu commented:
@noashin and @adrinjalali . Sorry for the second ping. Let me know if I should provide any changes!

adrinjalali reviewed (approved):
Thank you very much. This seems quite straightforward now, and with the two reviews we have, I'm confident with the contribution.
